gapis/replay/executor/executor.go

```diff
@@ -65,207 +60,85 @@ func Execute(
 	}.execute(ctx, connection)
 }
 
-func (e executor) execute(ctx context.Context, connection io.ReadWriteCloser) error {
-	// Encode the payload
-	// TODO: Make this a proto.
-	buf := &bytes.Buffer{}
-	w := endian.Writer(buf, e.memoryLayout.GetEndian())
-	w.Uint32(e.payload.StackSize)
-	w.Uint32(e.payload.VolatileMemorySize)
-	w.Uint32(uint32(len(e.payload.Constants)))
-	w.Data(e.payload.Constants)
-	w.Uint32(uint32(len(e.payload.Resources)))
-	for _, r := range e.payload.Resources {
-		w.String(r.ID)
-		w.Uint32(r.Size)
-	}
-	w.Uint32(uint32(len(e.payload.Opcodes)))
-	w.Data(e.payload.Opcodes)
-
-	data := buf.Bytes()
-
-	// Store the payload to the database
-	id, err := database.Store(ctx, data)
+func (e executor) execute(ctx context.Context, connection *gapir.Connection) error {
+	pp := e.payload.ToProto()
+	id, err := database.Store(ctx, &pp)
 	if err != nil {
-		return err
+		return log.Errf(ctx, err, "Storing replay payload")
 	}
 
 	// Kick the communication handler
-	responseR, responseW := io.Pipe()
-	comErr := make(chan error)
-	go func() {
-		err := e.handleReplayCommunication(ctx, connection, id, uint32(len(data)), responseW)
-		if err != nil {
-			log.W(ctx, "Replay communication failed: %v", err)
-			if closeErr := responseW.CloseWithError(err); closeErr != nil {
-				log.W(ctx, "Replay execute pipe writer CloseWithError: %v", closeErr)
-			}
-		} else {
-			if closeErr := responseW.Close(); closeErr != nil {
-				log.W(ctx, "Replay execute pipe writer Close failed: %v", closeErr)
-			}
-		}
-		comErr <- err
-	}()
-
-	// Decode and handle postbacks as they are received
-	e.decoder(responseR, nil)
-
-	err = <-comErr
-	if closeErr := responseR.Close(); closeErr != nil {
-		log.W(ctx, "Replay execute pipe reader Close failed: %v", closeErr)
-	}
+	err = connection.HandleReplayCommunication(
+		ctx, id.String(), e.payload,
+		e.handleResourceRequest,
```

Comment:
An alternative to N `funcs` here is to use an interface. Not arguing that it is a better design, but at least it would let you document the functions clearer.

Reply:
Changed to an "handler" interface.